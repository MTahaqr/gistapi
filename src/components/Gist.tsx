import React from 'react';

interface GistProps {
  gist: string;
}

const Gist = ({ gist }: GistProps) => <></>;

export default Gist;
