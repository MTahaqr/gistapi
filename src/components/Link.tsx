import React from 'react';
// Packages
import styled from 'styled-components';

interface LinkProps {
  href: string;
  target: string;
  rel?: string;
  children: React.ReactNode;
}

const Link: React.FC<LinkProps> = ({ href, target, rel = 'noreferrer', children }) => {
      /**
   * @Render
   */
  return (
    <LinkWrapper href={href} target={target} rel={rel}>
      {children}
    </LinkWrapper>
  );
};

export { Link, LinkProps };

const LinkWrapper = styled.a`
  cursor: pointer;
  text-decoration: none;
`;
