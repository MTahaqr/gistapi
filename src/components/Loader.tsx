import React from 'react';
// Packages
import { RotateSpinner } from 'react-spinners-kit';
import styled from 'styled-components';

interface LoaderProp {
  color?: string;
}
const Loader: React.FC<LoaderProp> = ({ color = '#24292e' }) => {
  /**
   * @Render
   */
  return (
    <StyledContainer>
      <RotateSpinner color={color} />
    </StyledContainer>
  );
};

const StyledContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
`;

export { Loader, LoaderProp };
