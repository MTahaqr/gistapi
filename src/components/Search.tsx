import React, { ChangeEvent, useEffect, useState } from 'react';
// Packages
import styled from 'styled-components';
// Hooks
import useDebounce from '../Hooks/useDebounce';
// Packages
import { AiOutlineSearch } from 'react-icons/ai';

interface SearchProps {
  onChange: (value: string) => void;
}
const Search = ({ onChange }: SearchProps) => {
  const [searchText, setSearchText] = useState<string>('');
  /**
   * @Methods
   */
  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    setSearchText(event.target.value || ''.trim());
  };

  const debouncedSearchVal = useDebounce(searchText, 500);

  /**
   * @Effects
   */
  useEffect(() => {
    onChange(debouncedSearchVal);
  }, [debouncedSearchVal, onChange]);

  /**
   * @Render
   */
  return (
    <Wrapper>
      <InputBox>
        <AiOutlineSearch color='black' size={20} />
        <Input
          value={searchText}
          placeholder='Search Gists for the username'
          onChange={handleChange}
        />
      </InputBox>
    </Wrapper>
  );
};

const Wrapper = styled.div`
  padding: 8px;
  background-color: #ffffff;
  font-size: 14px;
  line-height: 1.5;
  border-radius: 6px;
  margin: 0 16px;
`;

const InputBox = styled.div`
  border-radius: 4px;
  display: flex;
  width: 400px;
`;

const Input = styled.input`
  border: none;
  width: 100%;
  font-size: 16px;

  &:focus {
    outline: 0;
  }
`;

export { Search, SearchProps };
