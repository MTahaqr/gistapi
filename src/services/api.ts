import { QueryKey, useQuery, UseQueryOptions, UseQueryResult } from 'react-query';

export type ApiQueryOptions<Result, SelectResult = Result> = Omit<
  UseQueryOptions<Result, Error, SelectResult>,
  'queryFn' // queryFn is omitted because it is set at query hook creation
>;

export async function handleApiResult<T = unknown>(promise: Promise<{ data: T }>): Promise<T> {
  try {
    const { data } = await promise;
    return data as any;
  } catch (err: unknown) {
    alert(err);
    throw err;
  }
}

type ApiQueryHook<Params, Result> = <SelectResult = Result>(
  params: Params,
  options?: ApiQueryOptions<Result, SelectResult>,
) => UseQueryResult<SelectResult, Error>;

const buildApiQueryHook = <Params, Result>(
  cacheKey: string | ((params: Params) => QueryKey),
  fetch: (params: Params) => Promise<{ data: Result }>,
): ApiQueryHook<Params, Result> => {
  const useSpecificQuery = <SelectResult = Result>(
    params: Params,
    options?: ApiQueryOptions<Result, SelectResult>,
  ) => {
    return useQuery<Result, Error, SelectResult>(
      options?.queryKey || (typeof cacheKey === 'string' ? [cacheKey, params] : cacheKey(params)),
      () => handleApiResult<Result>(fetch(params)),
      {
        ...options,
        onError(error: Error) {
          if (options?.onError) options.onError(error);
        },
      },
    );
  };

  return useSpecificQuery;
};

export { buildApiQueryHook };
