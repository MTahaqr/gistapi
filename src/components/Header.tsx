import React from 'react';
import styled from 'styled-components';
// eslint-disable-next-line
// @ts-ignore skipping next line because no type was available for react-octicon library
// import Octicon from 'react-octicon';
import { Search } from './Search';
import { BsGithub } from 'react-icons/bs';

const Wrapper = styled.div`
  background-color: #24292e;
  position: sticky;
  top: 0px;
  width: 100%;
  color: #ffffff;
  z-index: 32;
  padding: 16px;
  font-size: 14px;
  line-height: 1.5;
  display: flex;
  align-items: center;
`;

interface HeaderProps {
  onChange: (value: string) => void;
}
const Header: React.FC<HeaderProps> = ({ onChange }) => {
  return (
    <Wrapper>
      <BsGithub size={32} />
      <Search onChange={onChange} />
    </Wrapper>
  );
};

export { Header, HeaderProps };
