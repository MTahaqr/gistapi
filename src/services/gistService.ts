import { Octokit } from '@octokit/rest';
import { buildApiQueryHook } from './api';
const octokit = new Octokit({ auth: process.env.REACT_APP_GITHUB_TOKEN });

export interface GistResult {
  url: string;
  forks_url: string;
  commits_url: string;
  id: string;
  comments_url: string;
  created_at: string;
  updated_at: string;
  description?: string | null;
  files: {
    [key: string]: {
      filename?: string | undefined;
      type?: string | undefined;
      language?: string | undefined;
      raw_url?: string | undefined;
      size?: number | undefined;
    };
  };
  owner?: {
    avatar_url: string;
    starred_url: string;
    login: string;
  };
}

interface usePublicGistsParams {
  perPage: number;
  pageNum: number;
}
interface useGistForUserParams {
  perPage: number;
  pageNum: number;
  userName: string;
}

export const getPublicGists = ({ perPage, pageNum }: usePublicGistsParams) =>
  octokit.gists.listPublic({ page: pageNum, per_page: perPage });

export const getGistForUser = ({
  userName,
  perPage,
  pageNum,
}: {
  userName: string;
  perPage: number;
  pageNum: number;
}) => octokit.gists.listForUser({ username: userName, page: pageNum, per_page: perPage });

export const usePublicGists = buildApiQueryHook<usePublicGistsParams, GistResult[]>(
  (params) => ['publicGist', params],
  (params) => getPublicGists({ ...params }),
);

export const useGistForUser = buildApiQueryHook<useGistForUserParams, GistResult[]>(
  (params) => ['gistForUser', params],
  (params) => getGistForUser({ ...params }),
);
