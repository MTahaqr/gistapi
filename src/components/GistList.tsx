import React, { useEffect, useState } from 'react';
import { GistResult } from '../services/gistService';
import styled from 'styled-components';
import { Gist } from './Gist';

const GistList = ({ gistList }: { gistList: GistResult[] }) => {
  return (
    <StyledContainer>
      {gistList?.map((data, index) => (
        <Gist key={index} {...data} />
      ))}
    </StyledContainer>
  );
};

export { GistList };

const StyledContainer = styled.div``;
