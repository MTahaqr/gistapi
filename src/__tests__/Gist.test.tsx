import { shallow, ShallowWrapper } from 'enzyme';
import * as React from 'react';
import { Gist } from '../components/Gist';
import { GistResult } from '../services/gistService';

describe('Gist', (): void => {
  let wrapper: ShallowWrapper;

  beforeEach((): void => {
    const props: GistResult = {
      url: 'https://api.github.com',
      forks_url: 'https://api.github.com',
      comments_url: 'https://api.github.com',
      commits_url: 'https://api.github.com',
      id: '232',
      files: {},
      owner: {
        login: 'https://api.github.com',
        avatar_url: 'https://api.github.com',
        starred_url: 'https://api.github.com',
      },
      description: 'https://api.github.com',
      created_at: '2023-01-31T08:15:12Z',
      updated_at: '2023-01-31T08:15:12Z',
    };
    wrapper = shallow(<Gist {...props} />);
  });

  it('should render properly', (): void => {
    expect(wrapper.getElement()).toMatchSnapshot();
  });
});
