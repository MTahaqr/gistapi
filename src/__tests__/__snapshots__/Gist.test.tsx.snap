// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gist should render properly 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      <styled.img
        alt="profile"
        src="https://api.github.com"
      />
      <styled.p>
        https://api.github.com
      </styled.p>
    </styled.div>
    <styled.div>
      <Link
        href="https://api.github.com"
        target="_blank"
      >
        <styled.div>
          <HiCodeBracket
            color="#0366D6"
            size={18}
          />
          <styled.p>
            0
             files
          </styled.p>
        </styled.div>
      </Link>
      <Link
        href="https://api.github.com"
        target="_blank"
      >
        <styled.div>
          <AiOutlineFork
            color="#0366D6"
            size={18}
          />
          <styled.p>
            forks
          </styled.p>
        </styled.div>
      </Link>
      <Link
        href="https://api.github.com"
        target="_blank"
      >
        <styled.div>
          <TfiComment
            color="#0366D6"
            size={18}
          />
          <styled.p>
            comments
          </styled.p>
        </styled.div>
      </Link>
      <Link
        href="https://api.github.com"
        target="_blank"
      >
        <styled.div>
          <AiFillStar
            color="#0366D6"
            size={18}
          />
          <styled.p>
            Stars
          </styled.p>
        </styled.div>
      </Link>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      Created at: 
      1/31/2023
    </styled.div>
    <styled.div>
      Last updated: 
      1/31/2023
    </styled.div>
  </styled.div>
  <styled.div>
    https://api.github.com
  </styled.div>
  <styled.div />
</styled.div>
`;
